core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1001 @@
+          trackersWithVolume.add(tracker);
+        }
+      }
+      return trackersWithVolume;
+    }
+
+    /**
+     * How many percent of the addresses would have their subchannels ejected if we proceeded
+     * with the next ejection.
+     */
+    double nextEjectionPercentage() {
+      int totalAddresses = 0;
+      int ejectedAddresses = 0;
+      for (AddressTracker tracker : trackerMap.values()) {
+        totalAddresses++;
+        if (tracker.subchannelsEjected()) {
+          ejectedAddresses++;
+        }
+      }
+      return ((double)(ejectedAddresses + 1) / totalAddresses) * 100;
+    }
+  }
+
+
+  /**
+   * Implementations provide different ways of ejecting outlier addresses..
+   */
+  interface OutlierEjectionAlgorithm {
+
+    /**
+     * Is the given {@link EquivalentAddressGroup} an outlier based on the past call results stored
+     * in {@link AddressTracker}.
+     */
+    void ejectOutliers(AddressTrackerMap trackerMap, long ejectionTimeMillis);
+
+    @Nullable
+    static OutlierEjectionAlgorithm forConfig(OutlierDetectionLoadBalancerConfig config) {
+      if (config.successRateEjection != null) {
+        return new SuccessRateOutlierEjectionAlgorithm(config);
+      } else if (config.failurePercentageEjection != null) {
+        return new FailurePercentageOutlierEjectionAlgorithm(config);
+      } else {
+        return null;
+      }
+    }
+  }
+
+  /**
+   * This algorithm ejects addresses that don't maintain a required rate of successful calls. The
+   * required rate is not fixed, but is based on the mean and standard deviation of the success
+   * rates of all of the addresses.
+   */
+  static class SuccessRateOutlierEjectionAlgorithm implements OutlierEjectionAlgorithm {
+
+    private final OutlierDetectionLoadBalancerConfig config;
+
+    SuccessRateOutlierEjectionAlgorithm(OutlierDetectionLoadBalancerConfig config) {
+      checkArgument(config.successRateEjection != null, "success rate ejection config is null");
+      this.config = config;
+    }
+
+    @Override
+    public void ejectOutliers(AddressTrackerMap trackerMap, long ejectionTimeNanos) {
+
+      // Only consider addresses that have the minimum request volume specified in the config.
+      List<AddressTracker> trackersWithVolume = trackerMap.trackersWithVolume(config);
+      // If we don't have enough addresses with significant volume then there's nothing to do.
+      if (trackersWithVolume.size() < config.successRateEjection.minimumHosts
+          || trackersWithVolume.size() == 0) {
+        return;
+      }
+
+      // Calculate mean and standard deviation of the fractions of successful calls.
+      List<Double> successRates = new ArrayList<>();
+      for (AddressTracker tracker : trackersWithVolume) {
+        successRates.add(tracker.successRate());
+      }
+      double mean = mean(successRates);
+      double stdev = standardDeviation(successRates, mean);
+
+      for (AddressTracker tracker : trackerMap.values()) {
```

Comment:
Only addresses with the minimum volume should be considered as candidates for ejection.

Reply:
I thought I was looping through the list of trackers that do have the minimum volume, but I wasn't. Fixed now and I also updated some the test for this as it was not sophisticated enough to catch this problem.